ICE: `open drop from non-ADT `

<!--
[31mICE[0m: Rustc ./a.rs '-Zvalidate-mir -Zinline-mir=yes -ooutputfile -Zdump-mir-dir=dir' 'error: internal compiler error: compiler/rustc_mir_transform/src/elaborate_drop.rs:1254:18: open drop from non-ADT `{type error}`', 'error: internal compiler error: compiler/rustc_mir_transform/src/elaborate_drop.rs:1254:18: open drop from non-ADT `{type error}`'
File: /tmp/im/a.rs
-->

snippet:
````rust
//@compile-flags: -Zvalidate-mir -Zinline-mir=yes
use std::mem::ManuallyDrop;
use std::{
    future::{async_drop_in_place, Future},
    pin::{pin, Pin},
};
fn a() {
    b(bar(0))
}
async fn bar(c: usize);
fn b<d>(e: d) {
    let mut f = pin!(ManuallyDrop::new(e));
    let mut fut = { Pin::map_unchecked_mut(f, |g| &mut **g) };
    let mut context;
    let h = async_drop_in_place(fut.get_unchecked_mut());
    let mut i = pin!(h);
    match i.poll( context) {
                      j }
}

````

Version information
````
rustc 1.89.0-nightly (718ddf660 2025-05-12)
binary: rustc
commit-hash: 718ddf660e6a1802c39b4962cf7eaa4db57025ef
commit-date: 2025-05-12
host: x86_64-unknown-linux-gnu
release: 1.89.0-nightly
LLVM version: 20.1.4

````

Possibly related line of code:
https://github.com/rust-lang/rust/blob/718ddf660e6a1802c39b4962cf7eaa4db57025ef/compiler/rustc_mir_transform/src/elaborate_drop.rs#L1248-L1260

Command:
`/home/matthias/.rustup/toolchains/master/bin/rustc -Zvalidate-mir -Zinline-mir=yes`

<details><summary><strong>Program output</strong></summary>
<p>

```
error[E0670]: `async fn` is not permitted in Rust 2015
 --> /tmp/icemaker_global_tempdir.tcPGNTRpBJ9s/rustc_testrunner_tmpdir_reporting.XOmIYrPbp6kH/mvce.rs:9:1
  |
9 | async fn bar(c: usize);
  | ^^^^^ to use `async fn`, switch to Rust 2018 or later
  |
  = help: pass `--edition 2024` to `rustc`
  = note: for more on editions, read https://doc.rust-lang.org/edition-guide

error: free function without a body
 --> /tmp/icemaker_global_tempdir.tcPGNTRpBJ9s/rustc_testrunner_tmpdir_reporting.XOmIYrPbp6kH/mvce.rs:9:1
  |
9 | async fn bar(c: usize);
  | ^^^^^^^^^^^^^^^^^^^^^^-
  |                       |
  |                       help: provide a definition for the function: `{ <body> }`

error: `match` arm with no body
  --> /tmp/icemaker_global_tempdir.tcPGNTRpBJ9s/rustc_testrunner_tmpdir_reporting.XOmIYrPbp6kH/mvce.rs:17:23
   |
17 |                       j }
   |                       ^
   |
help: add a body after the pattern
   |
17 |                       j => { todo!() } }
   |                         ++++++++++++++

warning: type parameter `d` should have an upper camel case name
  --> /tmp/icemaker_global_tempdir.tcPGNTRpBJ9s/rustc_testrunner_tmpdir_reporting.XOmIYrPbp6kH/mvce.rs:10:6
   |
10 | fn b<d>(e: d) {
   |      ^ help: convert the identifier to upper camel case: `D`
   |
   = note: `#[warn(non_camel_case_types)]` on by default

error[E0601]: `main` function not found in crate `mvce`
  --> /tmp/icemaker_global_tempdir.tcPGNTRpBJ9s/rustc_testrunner_tmpdir_reporting.XOmIYrPbp6kH/mvce.rs:18:2
   |
18 | }
   |  ^ consider adding a `main` function to `/tmp/icemaker_global_tempdir.tcPGNTRpBJ9s/rustc_testrunner_tmpdir_reporting.XOmIYrPbp6kH/mvce.rs`

error[E0658]: use of unstable library feature `async_drop`
 --> /tmp/icemaker_global_tempdir.tcPGNTRpBJ9s/rustc_testrunner_tmpdir_reporting.XOmIYrPbp6kH/mvce.rs:3:14
  |
3 |     future::{async_drop_in_place, Future},
  |              ^^^^^^^^^^^^^^^^^^^
  |
  = note: see issue #126482 <https://github.com/rust-lang/rust/issues/126482> for more information
  = help: add `#![feature(async_drop)]` to the crate attributes to enable
  = note: this compiler was built on 2025-05-12; consider upgrading it if it is out of date

error[E0658]: use of unstable library feature `async_drop`
  --> /tmp/icemaker_global_tempdir.tcPGNTRpBJ9s/rustc_testrunner_tmpdir_reporting.XOmIYrPbp6kH/mvce.rs:14:13
   |
14 |     let h = async_drop_in_place(fut.get_unchecked_mut());
   |             ^^^^^^^^^^^^^^^^^^^
   |
   = note: see issue #126482 <https://github.com/rust-lang/rust/issues/126482> for more information
   = help: add `#![feature(async_drop)]` to the crate attributes to enable
   = note: this compiler was built on 2025-05-12; consider upgrading it if it is out of date

warning: unused variable: `j`
  --> /tmp/icemaker_global_tempdir.tcPGNTRpBJ9s/rustc_testrunner_tmpdir_reporting.XOmIYrPbp6kH/mvce.rs:17:23
   |
17 |                       j }
   |                       ^ help: if this is intentional, prefix it with an underscore: `_j`
   |
   = note: `#[warn(unused_variables)]` on by default

error[E0133]: call to unsafe function `Pin::<&'a mut T>::map_unchecked_mut` is unsafe and requires unsafe function or block
  --> /tmp/icemaker_global_tempdir.tcPGNTRpBJ9s/rustc_testrunner_tmpdir_reporting.XOmIYrPbp6kH/mvce.rs:12:21
   |
12 |     let mut fut = { Pin::map_unchecked_mut(f, |g| &mut **g) };
   |                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ call to unsafe function
   |
   = note: consult the function's documentation for information on how to avoid undefined behavior

error[E0133]: call to unsafe function `async_drop_in_place` is unsafe and requires unsafe function or block
  --> /tmp/icemaker_global_tempdir.tcPGNTRpBJ9s/rustc_testrunner_tmpdir_reporting.XOmIYrPbp6kH/mvce.rs:14:13
   |
14 |     let h = async_drop_in_place(fut.get_unchecked_mut());
   |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ call to unsafe function
   |
   = note: consult the function's documentation for information on how to avoid undefined behavior

error[E0133]: call to unsafe function `Pin::<&'a mut T>::get_unchecked_mut` is unsafe and requires unsafe function or block
  --> /tmp/icemaker_global_tempdir.tcPGNTRpBJ9s/rustc_testrunner_tmpdir_reporting.XOmIYrPbp6kH/mvce.rs:14:33
   |
14 |     let h = async_drop_in_place(fut.get_unchecked_mut());
   |                                 ^^^^^^^^^^^^^^^^^^^^^^^ call to unsafe function
   |
   = note: consult the function's documentation for information on how to avoid undefined behavior

warning: variable does not need to be mutable
  --> /tmp/icemaker_global_tempdir.tcPGNTRpBJ9s/rustc_testrunner_tmpdir_reporting.XOmIYrPbp6kH/mvce.rs:11:9
   |
11 |     let mut f = pin!(ManuallyDrop::new(e));
   |         ----^
   |         |
   |         help: remove this `mut`
   |
   = note: `#[warn(unused_mut)]` on by default

warning: variable does not need to be mutable
  --> /tmp/icemaker_global_tempdir.tcPGNTRpBJ9s/rustc_testrunner_tmpdir_reporting.XOmIYrPbp6kH/mvce.rs:12:9
   |
12 |     let mut fut = { Pin::map_unchecked_mut(f, |g| &mut **g) };
   |         ----^^^
   |         |
   |         help: remove this `mut`

warning: variable does not need to be mutable
  --> /tmp/icemaker_global_tempdir.tcPGNTRpBJ9s/rustc_testrunner_tmpdir_reporting.XOmIYrPbp6kH/mvce.rs:15:9
   |
15 |     let mut i = pin!(h);
   |         ----^
   |         |
   |         help: remove this `mut`

error[E0381]: used binding `context` isn't initialized
  --> /tmp/icemaker_global_tempdir.tcPGNTRpBJ9s/rustc_testrunner_tmpdir_reporting.XOmIYrPbp6kH/mvce.rs:16:19
   |
13 |     let mut context;
   |         ----------- binding declared here but left uninitialized
...
16 |     match i.poll( context) {
   |                   ^^^^^^^ `context` used here but it isn't initialized
   |
help: consider assigning a value
   |
13 |     let mut context = &mut /* value */;
   |                     ++++++++++++++++++

error: internal compiler error: compiler/rustc_mir_transform/src/elaborate_drop.rs:1254:18: open drop from non-ADT `{type error}`
  --> /home/matthias/.rustup/toolchains/master/lib/rustlib/src/rust/library/core/src/future/async_drop.rs:48:70
   |
48 |   pub async unsafe fn async_drop_in_place<T: ?Sized>(_to_drop: *mut T) {
   |  ______________________________________________________________________^
...  |
51 | | }
   | |_^


thread 'rustc' panicked at compiler/rustc_mir_transform/src/elaborate_drop.rs:1254:18:
Box<dyn Any>
stack backtrace:
   0:     0x78a0ab959fb3 - <std::sys::backtrace::BacktraceLock::print::DisplayBacktrace as core::fmt::Display>::fmt::hd03fa4624d6a43ff
   1:     0x78a0ac005987 - core::fmt::write::h99c6bd700e41fe5b
   2:     0x78a0ad0b4091 - std::io::Write::write_fmt::h8fdba04fd8ac1437
   3:     0x78a0ab959e12 - std::sys::backtrace::BacktraceLock::print::ha37796b30caa3dee
   4:     0x78a0ab95da0a - std::panicking::default_hook::{{closure}}::h2be5a8e7cd709dcc
   5:     0x78a0ab95d58f - std::panicking::default_hook::h652a4375123f2a56
   6:     0x78a0aa98c873 - std[56b07bdfd6b41829]::panicking::update_hook::<alloc[763664f815bf0fa8]::boxed::Box<rustc_driver_impl[fb64ae308f02562]::install_ice_hook::{closure#1}>>::{closure#0}
   7:     0x78a0ab95e283 - std::panicking::rust_panic_with_hook::h4f131a958fbdba61
   8:     0x78a0aa9c85f1 - std[56b07bdfd6b41829]::panicking::begin_panic::<rustc_errors[b8a92ed9b61c1989]::ExplicitBug>::{closure#0}
   9:     0x78a0aa9bc626 - std[56b07bdfd6b41829]::sys::backtrace::__rust_end_short_backtrace::<std[56b07bdfd6b41829]::panicking::begin_panic<rustc_errors[b8a92ed9b61c1989]::ExplicitBug>::{closure#0}, !>
  10:     0x78a0aa9b8ce9 - std[56b07bdfd6b41829]::panicking::begin_panic::<rustc_errors[b8a92ed9b61c1989]::ExplicitBug>
  11:     0x78a0aa9d2c71 - <rustc_errors[b8a92ed9b61c1989]::diagnostic::BugAbort as rustc_errors[b8a92ed9b61c1989]::diagnostic::EmissionGuarantee>::emit_producing_guarantee
  12:     0x78a0aafa0adc - <rustc_errors[b8a92ed9b61c1989]::DiagCtxtHandle>::span_bug::<rustc_span[d026ac1b2b8daa7c]::span_encoding::Span, alloc[763664f815bf0fa8]::string::String>
  13:     0x78a0ab03ecb7 - rustc_middle[60b471758888797e]::util::bug::opt_span_bug_fmt::<rustc_span[d026ac1b2b8daa7c]::span_encoding::Span>::{closure#0}
  14:     0x78a0ab01723a - rustc_middle[60b471758888797e]::ty::context::tls::with_opt::<rustc_middle[60b471758888797e]::util::bug::opt_span_bug_fmt<rustc_span[d026ac1b2b8daa7c]::span_encoding::Span>::{closure#0}, !>::{closure#0}
  15:     0x78a0ab0170ab - rustc_middle[60b471758888797e]::ty::context::tls::with_context_opt::<rustc_middle[60b471758888797e]::ty::context::tls::with_opt<rustc_middle[60b471758888797e]::util::bug::opt_span_bug_fmt<rustc_span[d026ac1b2b8daa7c]::span_encoding::Span>::{closure#0}, !>::{closure#0}, !>
  16:     0x78a0a93f14c7 - rustc_middle[60b471758888797e]::util::bug::span_bug_fmt::<rustc_span[d026ac1b2b8daa7c]::span_encoding::Span>
  17:     0x78a0ac897c24 - <rustc_mir_transform[4a4e9cf23a9606fd]::elaborate_drop::DropCtxt<rustc_mir_transform[4a4e9cf23a9606fd]::shim::DropShimElaborator>>::elaborate_drop
  18:     0x78a0ac891851 - rustc_mir_transform[4a4e9cf23a9606fd]::shim::make_shim
  19:     0x78a0ac88f82f - rustc_query_impl[cf6f12f07b5d9141]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[cf6f12f07b5d9141]::query_impl::mir_shims::dynamic_query::{closure#2}::{closure#0}, rustc_middle[60b471758888797e]::query::erase::Erased<[u8; 8usize]>>
  20:     0x78a0ac88f7e3 - <rustc_query_impl[cf6f12f07b5d9141]::query_impl::mir_shims::dynamic_query::{closure#2} as core[604f71d46da0bdba]::ops::function::FnOnce<(rustc_middle[60b471758888797e]::ty::context::TyCtxt, rustc_middle[60b471758888797e]::ty::instance::InstanceKind)>>::call_once
  21:     0x78a0ac852d8c - rustc_query_system[620f1d082c8e701b]::query::plumbing::try_execute_query::<rustc_query_impl[cf6f12f07b5d9141]::DynamicConfig<rustc_query_system[620f1d082c8e701b]::query::caches::DefaultCache<rustc_middle[60b471758888797e]::ty::instance::InstanceKind, rustc_middle[60b471758888797e]::query::erase::Erased<[u8; 8usize]>>, false, false, false>, rustc_query_impl[cf6f12f07b5d9141]::plumbing::QueryCtxt, false>
  22:     0x78a0ac852aaa - rustc_query_impl[cf6f12f07b5d9141]::query_impl::mir_shims::get_query_non_incr::__rust_end_short_backtrace
  23:     0x78a0ac84d95e - <rustc_middle[60b471758888797e]::ty::context::TyCtxt>::instance_mir
  24:     0x78a0aca7a70f - rustc_mir_transform[4a4e9cf23a9606fd]::inline::cycle::mir_inliner_callees
  25:     0x78a0aca7a574 - rustc_query_impl[cf6f12f07b5d9141]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[cf6f12f07b5d9141]::query_impl::mir_inliner_callees::dynamic_query::{closure#2}::{closure#0}, rustc_middle[60b471758888797e]::query::erase::Erased<[u8; 16usize]>>
  26:     0x78a0aca7a543 - <rustc_query_impl[cf6f12f07b5d9141]::query_impl::mir_inliner_callees::dynamic_query::{closure#2} as core[604f71d46da0bdba]::ops::function::FnOnce<(rustc_middle[60b471758888797e]::ty::context::TyCtxt, rustc_middle[60b471758888797e]::ty::instance::InstanceKind)>>::call_once
  27:     0x78a0ac853a47 - rustc_query_system[620f1d082c8e701b]::query::plumbing::try_execute_query::<rustc_query_impl[cf6f12f07b5d9141]::DynamicConfig<rustc_query_system[620f1d082c8e701b]::query::caches::DefaultCache<rustc_middle[60b471758888797e]::ty::instance::InstanceKind, rustc_middle[60b471758888797e]::query::erase::Erased<[u8; 16usize]>>, false, false, false>, rustc_query_impl[cf6f12f07b5d9141]::plumbing::QueryCtxt, false>
  28:     0x78a0ac85373b - rustc_query_impl[cf6f12f07b5d9141]::query_impl::mir_inliner_callees::get_query_non_incr::__rust_end_short_backtrace
  29:     0x78a0acff4d76 - rustc_mir_transform[4a4e9cf23a9606fd]::inline::cycle::mir_callgraph_reachable::process
  30:     0x78a0acff519f - rustc_mir_transform[4a4e9cf23a9606fd]::inline::cycle::mir_callgraph_reachable::process
  31:     0x78a0acff4544 - rustc_mir_transform[4a4e9cf23a9606fd]::inline::cycle::mir_callgraph_reachable
  32:     0x78a0acff4369 - rustc_query_impl[cf6f12f07b5d9141]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[cf6f12f07b5d9141]::query_impl::mir_callgraph_reachable::dynamic_query::{closure#2}::{closure#0}, rustc_middle[60b471758888797e]::query::erase::Erased<[u8; 1usize]>>
  33:     0x78a0acff4339 - <rustc_query_impl[cf6f12f07b5d9141]::query_impl::mir_callgraph_reachable::dynamic_query::{closure#2} as core[604f71d46da0bdba]::ops::function::FnOnce<(rustc_middle[60b471758888797e]::ty::context::TyCtxt, (rustc_middle[60b471758888797e]::ty::instance::Instance, rustc_span[d026ac1b2b8daa7c]::def_id::LocalDefId))>>::call_once
  34:     0x78a0ac851dac - rustc_query_system[620f1d082c8e701b]::query::plumbing::try_execute_query::<rustc_query_impl[cf6f12f07b5d9141]::DynamicConfig<rustc_query_system[620f1d082c8e701b]::query::caches::DefaultCache<(rustc_middle[60b471758888797e]::ty::instance::Instance, rustc_span[d026ac1b2b8daa7c]::def_id::LocalDefId), rustc_middle[60b471758888797e]::query::erase::Erased<[u8; 1usize]>>, false, false, false>, rustc_query_impl[cf6f12f07b5d9141]::plumbing::QueryCtxt, false>
  35:     0x78a0ac851aa8 - rustc_query_impl[cf6f12f07b5d9141]::query_impl::mir_callgraph_reachable::get_query_non_incr::__rust_end_short_backtrace
  36:     0x78a0ac8a2a7b - rustc_mir_transform[4a4e9cf23a9606fd]::inline::process_blocks::<rustc_mir_transform[4a4e9cf23a9606fd]::inline::NormalInliner>
  37:     0x78a0ac8cd31a - <rustc_mir_transform[4a4e9cf23a9606fd]::inline::Inline as rustc_mir_transform[4a4e9cf23a9606fd]::pass_manager::MirPass>::run_pass
  38:     0x78a0ac00383d - rustc_mir_transform[4a4e9cf23a9606fd]::pass_manager::run_passes_inner
  39:     0x78a0aca7da03 - rustc_mir_transform[4a4e9cf23a9606fd]::run_optimization_passes
  40:     0x78a0aca7b2fb - rustc_mir_transform[4a4e9cf23a9606fd]::optimized_mir
  41:     0x78a0aca7ae9d - rustc_query_impl[cf6f12f07b5d9141]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[cf6f12f07b5d9141]::query_impl::optimized_mir::dynamic_query::{closure#2}::{closure#0}, rustc_middle[60b471758888797e]::query::erase::Erased<[u8; 8usize]>>
  42:     0x78a0ac450e3f - rustc_query_system[620f1d082c8e701b]::query::plumbing::try_execute_query::<rustc_query_impl[cf6f12f07b5d9141]::DynamicConfig<rustc_query_system[620f1d082c8e701b]::query::caches::DefIdCache<rustc_middle[60b471758888797e]::query::erase::Erased<[u8; 8usize]>>, false, false, false>, rustc_query_impl[cf6f12f07b5d9141]::plumbing::QueryCtxt, false>
  43:     0x78a0ac450262 - rustc_query_impl[cf6f12f07b5d9141]::query_impl::optimized_mir::get_query_non_incr::__rust_end_short_backtrace
  44:     0x78a0ac84d286 - <rustc_middle[60b471758888797e]::ty::context::TyCtxt>::instance_mir
  45:     0x78a0ac4b585f - rustc_interface[52ab4bdc9c3114ed]::passes::run_required_analyses
  46:     0x78a0acfbb49e - rustc_interface[52ab4bdc9c3114ed]::passes::analysis
  47:     0x78a0acfbb475 - rustc_query_impl[cf6f12f07b5d9141]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[cf6f12f07b5d9141]::query_impl::analysis::dynamic_query::{closure#2}::{closure#0}, rustc_middle[60b471758888797e]::query::erase::Erased<[u8; 0usize]>>
  48:     0x78a0acf2cdba - rustc_query_system[620f1d082c8e701b]::query::plumbing::try_execute_query::<rustc_query_impl[cf6f12f07b5d9141]::DynamicConfig<rustc_query_system[620f1d082c8e701b]::query::caches::SingleCache<rustc_middle[60b471758888797e]::query::erase::Erased<[u8; 0usize]>>, false, false, false>, rustc_query_impl[cf6f12f07b5d9141]::plumbing::QueryCtxt, false>
  49:     0x78a0acf2ca8f - rustc_query_impl[cf6f12f07b5d9141]::query_impl::analysis::get_query_non_incr::__rust_end_short_backtrace
  50:     0x78a0ad1fdc07 - rustc_interface[52ab4bdc9c3114ed]::passes::create_and_enter_global_ctxt::<core[604f71d46da0bdba]::option::Option<rustc_interface[52ab4bdc9c3114ed]::queries::Linker>, rustc_driver_impl[fb64ae308f02562]::run_compiler::{closure#0}::{closure#2}>::{closure#2}::{closure#0}
  51:     0x78a0ad0f9fe6 - rustc_interface[52ab4bdc9c3114ed]::interface::run_compiler::<(), rustc_driver_impl[fb64ae308f02562]::run_compiler::{closure#0}>::{closure#1}
  52:     0x78a0ad0aacfe - std[56b07bdfd6b41829]::sys::backtrace::__rust_begin_short_backtrace::<rustc_interface[52ab4bdc9c3114ed]::util::run_in_thread_with_globals<rustc_interface[52ab4bdc9c3114ed]::util::run_in_thread_pool_with_globals<rustc_interface[52ab4bdc9c3114ed]::interface::run_compiler<(), rustc_driver_impl[fb64ae308f02562]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>
  53:     0x78a0ad0ab16b - <<std[56b07bdfd6b41829]::thread::Builder>::spawn_unchecked_<rustc_interface[52ab4bdc9c3114ed]::util::run_in_thread_with_globals<rustc_interface[52ab4bdc9c3114ed]::util::run_in_thread_pool_with_globals<rustc_interface[52ab4bdc9c3114ed]::interface::run_compiler<(), rustc_driver_impl[fb64ae308f02562]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>::{closure#1} as core[604f71d46da0bdba]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  54:     0x78a0ad0ac56b - std::sys::pal::unix::thread::Thread::new::thread_start::he52776061f67c85f
  55:     0x78a0a6ea370a - <unknown>
  56:     0x78a0a6f27aac - <unknown>
  57:                0x0 - <unknown>

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: please make sure that you have updated to the latest nightly

note: rustc 1.89.0-nightly (718ddf660 2025-05-12) running on x86_64-unknown-linux-gnu

note: compiler flags: -Z validate-mir -Z inline-mir=yes -Z dump-mir-dir=dir

query stack during panic:
#0 [mir_shims] generating MIR shim for `core::future::async_drop::async_drop_in_place::{closure#0}`, instance=AsyncDropGlue(DefId(2:15267 ~ core[604f]::future::async_drop::async_drop_in_place::{closure#0}), Coroutine(DefId(2:15267 ~ core[604f]::future::async_drop::async_drop_in_place::{closure#0}), [{type error}, (), core::future::ResumeTy, (), (), CoroutineWitness(DefId(2:15267 ~ core[604f]::future::async_drop::async_drop_in_place::{closure#0}), [{type error}]), (*mut {type error},)]))
#1 [mir_inliner_callees] computing all local function calls in `core::future::async_drop::async_drop_in_place::{closure#0}`
#2 [mir_callgraph_reachable] computing if `b::<{type error}>` (transitively) calls `a`
#3 [optimized_mir] optimizing MIR for `a`
#4 [analysis] running analysis passes on this crate
end of query stack
error: aborting due to 11 previous errors; 5 warnings emitted

Some errors have detailed explanations: E0133, E0381, E0601, E0658, E0670.
For more information about an error, try `rustc --explain E0133`.

```

</p>
</details>

<!--
query stack:
#0 [mir_shims] generating MIR shim for `core::future::async_drop::async_drop_in_place::{closure#0}`, instance=AsyncDropGlue(DefId(2:15267 ~ core[604f]::future::async_drop::async_drop_in_place::{closure#0}), Coroutine(DefId(2:15267 ~ core[604f]::future::async_drop::async_drop_in_place::{closure#0}), [{type error}, (), core::future::ResumeTy, (), (), CoroutineWitness(DefId(2:15267 ~ core[604f]::future::async_drop::async_drop_in_place::{closure#0}), [{type error}]), (*mut {type error},)]))
#1 [mir_inliner_callees] computing all local function calls in `core::future::async_drop::async_drop_in_place::{closure#0}`
#2 [mir_callgraph_reachable] computing if `b::<{type error}>` (transitively) calls `a`
#3 [optimized_mir] optimizing MIR for `a`
#4 [analysis] running analysis passes on this crate

-->

Minimized:
```rust
//@compile-flags: -Zvalidate-mir -Zinline-mir=yes --crate-type=lib

#![feature(async_drop)]

use std::{
    future::{Future, async_drop_in_place},
    pin::pin,
    task::Context,
};

fn wrong() -> impl Sized {
    *"abc" // Doesn't implement Sized
}
fn weird(context: &mut Context<'_>) {
    let mut e = wrong();
    let h = unsafe { async_drop_in_place(&raw mut e) };
    let i = pin!(h);
    i.poll(context);
}
```

<details>
<summary>Error output</summary>

```
warning: the feature `async_drop` is incomplete and may not be safe to use and/or cause compiler crashes
 --> src/lib.rs:3:12
  |
3 | #![feature(async_drop)]
  |            ^^^^^^^^^^
  |
  = note: see issue #126482 <https://github.com/rust-lang/rust/issues/126482> for more information
  = note: `#[warn(incomplete_features)]` on by default

error[E0277]: the size for values of type `str` cannot be known at compilation time
  --> src/lib.rs:11:15
   |
11 | fn wrong() -> impl Sized {
   |               ^^^^^^^^^^ doesn't have a size known at compile-time
12 |     *"abc" // Doesn't implement Sized
   |     ------ return type was inferred to be `str` here
   |
   = help: the trait `Sized` is not implemented for `str`
help: references are always `Sized`, even if they point to unsized data; consider not dereferencing the expression
   |
12 -     *"abc" // Doesn't implement Sized
12 +     "abc" // Doesn't implement Sized
   |

error: internal compiler error: compiler/rustc_mir_transform/src/elaborate_drop.rs:1254:18: open drop from non-ADT `{type error}`
  --> /Users/timch/.rustup/toolchains/nightly-aarch64-apple-darwin/lib/rustlib/src/rust/library/core/src/future/async_drop.rs:48:70
   |
48 |   pub async unsafe fn async_drop_in_place<T: ?Sized>(_to_drop: *mut T) {
   |  ______________________________________________________________________^
...  |
51 | | }
   | |_^


thread 'rustc' panicked at compiler/rustc_mir_transform/src/elaborate_drop.rs:1254:18:
Box<dyn Any>
stack backtrace:
   0:        0x1103ca290 - <std::sys::backtrace::BacktraceLock::print::DisplayBacktrace as core::fmt::Display>::fmt::h8538a0cbd0a29488
   1:        0x10d9b91e8 - core::fmt::write::hc868a20baa39afeb
   2:        0x1103bebc4 - std::io::Write::write_fmt::h3d4b5fdc963d14ff
   3:        0x1103ca150 - std::sys::backtrace::BacktraceLock::print::h93cc182a120c41ba
   4:        0x1103cee0c - std::panicking::default_hook::{{closure}}::hb672fcbc7dc2d6ae
   5:        0x1103ceadc - std::panicking::default_hook::h2c53d6eae91798f2
   6:        0x10e527db8 - std[37ce58b9063475e]::panicking::update_hook::<alloc[b1849f567b850016]::boxed::Box<rustc_driver_impl[f8d7c4f8b6e94698]::install_ice_hook::{closure#1}>>::{closure#0}
   7:        0x1103cf7f4 - std::panicking::rust_panic_with_hook::hc7af3aa4fdc12c97
   8:        0x10e5aacd4 - std[37ce58b9063475e]::panicking::begin_panic::<rustc_errors[e3d83480659768f1]::ExplicitBug>::{closure#0}
   9:        0x10e598e80 - std[37ce58b9063475e]::sys::backtrace::__rust_end_short_backtrace::<std[37ce58b9063475e]::panicking::begin_panic<rustc_errors[e3d83480659768f1]::ExplicitBug>::{closure#0}, !>
  10:        0x1130651ac - std[37ce58b9063475e]::panicking::begin_panic::<rustc_errors[e3d83480659768f1]::ExplicitBug>
  11:        0x113065584 - <rustc_errors[e3d83480659768f1]::diagnostic::BugAbort as rustc_errors[e3d83480659768f1]::diagnostic::EmissionGuarantee>::emit_producing_guarantee
  12:        0x1130d5f98 - <rustc_errors[e3d83480659768f1]::DiagCtxtHandle>::span_bug::<rustc_span[594e537fab3ea3f]::span_encoding::Span, alloc[b1849f567b850016]::string::String>
  13:        0x1130dee1c - rustc_middle[e7ee0c50ed9ed64b]::util::bug::opt_span_bug_fmt::<rustc_span[594e537fab3ea3f]::span_encoding::Span>::{closure#0}
  14:        0x10f19eb24 - rustc_middle[e7ee0c50ed9ed64b]::ty::context::tls::with_opt::<rustc_middle[e7ee0c50ed9ed64b]::util::bug::opt_span_bug_fmt<rustc_span[594e537fab3ea3f]::span_encoding::Span>::{closure#0}, !>::{closure#0}
  15:        0x10f19e5f4 - rustc_middle[e7ee0c50ed9ed64b]::ty::context::tls::with_context_opt::<rustc_middle[e7ee0c50ed9ed64b]::ty::context::tls::with_opt<rustc_middle[e7ee0c50ed9ed64b]::util::bug::opt_span_bug_fmt<rustc_span[594e537fab3ea3f]::span_encoding::Span>::{closure#0}, !>::{closure#0}, !>
  16:        0x1130ded8c - rustc_middle[e7ee0c50ed9ed64b]::util::bug::span_bug_fmt::<rustc_span[594e537fab3ea3f]::span_encoding::Span>
  17:        0x10f4a12d0 - <rustc_mir_transform[1f3f7b4c1c2588bc]::elaborate_drop::DropCtxt<rustc_mir_transform[1f3f7b4c1c2588bc]::shim::DropShimElaborator>>::elaborate_drop
  18:        0x10f493e80 - rustc_mir_transform[1f3f7b4c1c2588bc]::shim::make_shim
  19:        0x10fb273ac - rustc_query_impl[450b0d7dafedda2d]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[450b0d7dafedda2d]::query_impl::mir_shims::dynamic_query::{closure#2}::{closure#0}, rustc_middle[e7ee0c50ed9ed64b]::query::erase::Erased<[u8; 8usize]>>
  20:        0x10fafbba0 - <rustc_query_impl[450b0d7dafedda2d]::query_impl::mir_shims::dynamic_query::{closure#2} as core[8358e3f6e6ac61e1]::ops::function::FnOnce<(rustc_middle[e7ee0c50ed9ed64b]::ty::context::TyCtxt, rustc_middle[e7ee0c50ed9ed64b]::ty::instance::InstanceKind)>>::call_once
  21:        0x10f942cac - rustc_query_system[f04a75388b1923fa]::query::plumbing::try_execute_query::<rustc_query_impl[450b0d7dafedda2d]::DynamicConfig<rustc_query_system[f04a75388b1923fa]::query::caches::DefaultCache<rustc_middle[e7ee0c50ed9ed64b]::ty::instance::InstanceKind, rustc_middle[e7ee0c50ed9ed64b]::query::erase::Erased<[u8; 8usize]>>, false, false, false>, rustc_query_impl[450b0d7dafedda2d]::plumbing::QueryCtxt, false>
  22:        0x10fb63ca4 - rustc_query_impl[450b0d7dafedda2d]::query_impl::mir_shims::get_query_non_incr::__rust_end_short_backtrace
  23:        0x10f223a10 - <rustc_middle[e7ee0c50ed9ed64b]::ty::context::TyCtxt>::instance_mir
  24:        0x10f544750 - rustc_mir_transform[1f3f7b4c1c2588bc]::inline::try_instance_mir
  25:        0x10f4f65f4 - rustc_mir_transform[1f3f7b4c1c2588bc]::inline::process_blocks::<rustc_mir_transform[1f3f7b4c1c2588bc]::inline::NormalInliner>
  26:        0x10f4f234c - <rustc_mir_transform[1f3f7b4c1c2588bc]::inline::Inline as rustc_mir_transform[1f3f7b4c1c2588bc]::pass_manager::MirPass>::run_pass
  27:        0x10f4795e0 - rustc_mir_transform[1f3f7b4c1c2588bc]::pass_manager::run_passes_inner
  28:        0x10f49e5cc - rustc_mir_transform[1f3f7b4c1c2588bc]::run_optimization_passes
  29:        0x10f49ea4c - rustc_mir_transform[1f3f7b4c1c2588bc]::optimized_mir
  30:        0x10fb2195c - rustc_query_impl[450b0d7dafedda2d]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[450b0d7dafedda2d]::query_impl::optimized_mir::dynamic_query::{closure#2}::{closure#0}, rustc_middle[e7ee0c50ed9ed64b]::query::erase::Erased<[u8; 8usize]>>
  31:        0x10f8f4334 - rustc_query_system[f04a75388b1923fa]::query::plumbing::try_execute_query::<rustc_query_impl[450b0d7dafedda2d]::DynamicConfig<rustc_query_system[f04a75388b1923fa]::query::caches::DefIdCache<rustc_middle[e7ee0c50ed9ed64b]::query::erase::Erased<[u8; 8usize]>>, false, false, false>, rustc_query_impl[450b0d7dafedda2d]::plumbing::QueryCtxt, false>
  32:        0x10fb3eb98 - rustc_query_impl[450b0d7dafedda2d]::query_impl::optimized_mir::get_query_non_incr::__rust_end_short_backtrace
  33:        0x10f223c3c - <rustc_middle[e7ee0c50ed9ed64b]::ty::context::TyCtxt>::instance_mir
  34:        0x10ee0f884 - rustc_interface[70e180df94007202]::passes::run_required_analyses
  35:        0x10ee112e8 - rustc_interface[70e180df94007202]::passes::analysis
  36:        0x10fb26fd4 - rustc_query_impl[450b0d7dafedda2d]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[450b0d7dafedda2d]::query_impl::analysis::dynamic_query::{closure#2}::{closure#0}, rustc_middle[e7ee0c50ed9ed64b]::query::erase::Erased<[u8; 0usize]>>
  37:        0x10f8f7fe8 - rustc_query_system[f04a75388b1923fa]::query::plumbing::try_execute_query::<rustc_query_impl[450b0d7dafedda2d]::DynamicConfig<rustc_query_system[f04a75388b1923fa]::query::caches::SingleCache<rustc_middle[e7ee0c50ed9ed64b]::query::erase::Erased<[u8; 0usize]>>, false, false, false>, rustc_query_impl[450b0d7dafedda2d]::plumbing::QueryCtxt, false>
  38:        0x10fb33258 - rustc_query_impl[450b0d7dafedda2d]::query_impl::analysis::get_query_non_incr::__rust_end_short_backtrace
  39:        0x10e4d8dfc - rustc_interface[70e180df94007202]::passes::create_and_enter_global_ctxt::<core[8358e3f6e6ac61e1]::option::Option<rustc_interface[70e180df94007202]::queries::Linker>, rustc_driver_impl[f8d7c4f8b6e94698]::run_compiler::{closure#0}::{closure#2}>
  40:        0x10e524bb8 - rustc_interface[70e180df94007202]::interface::run_compiler::<(), rustc_driver_impl[f8d7c4f8b6e94698]::run_compiler::{closure#0}>::{closure#1}
  41:        0x10e515b10 - std[37ce58b9063475e]::sys::backtrace::__rust_begin_short_backtrace::<rustc_interface[70e180df94007202]::util::run_in_thread_with_globals<rustc_interface[70e180df94007202]::util::run_in_thread_pool_with_globals<rustc_interface[70e180df94007202]::interface::run_compiler<(), rustc_driver_impl[f8d7c4f8b6e94698]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>
  42:        0x10e52b574 - <<std[37ce58b9063475e]::thread::Builder>::spawn_unchecked_<rustc_interface[70e180df94007202]::util::run_in_thread_with_globals<rustc_interface[70e180df94007202]::util::run_in_thread_pool_with_globals<rustc_interface[70e180df94007202]::interface::run_compiler<(), rustc_driver_impl[f8d7c4f8b6e94698]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>::{closure#1} as core[8358e3f6e6ac61e1]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  43:        0x1103d1d38 - std::sys::pal::unix::thread::Thread::new::thread_start::h1a4d34a906f31054
  44:        0x190301c0c - __pthread_cond_wait

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: please make sure that you have updated to the latest nightly

note: please attach the file at `/Users/timch/foo/rustc-ice-2025-05-12T07_30_26-11777.txt` to your bug report

note: compiler flags: -Z validate-mir -Z inline-mir=yes --crate-type lib

query stack during panic:
#0 [mir_shims] generating MIR shim for `core::future::async_drop::async_drop_in_place::{closure#0}`, instance=AsyncDropGlue(DefId(2:15263 ~ core[8358]::future::async_drop::async_drop_in_place::{closure#0}), Coroutine(DefId(2:15263 ~ core[8358]::future::async_drop::async_drop_in_place::{closure#0}), [{type error}, (), core::future::ResumeTy, (), (), CoroutineWitness(DefId(2:15263 ~ core[8358]::future::async_drop::async_drop_in_place::{closure#0}), [{type error}]), (*mut {type error},)]))
#1 [optimized_mir] optimizing MIR for `weird`
#2 [analysis] running analysis passes on this crate
end of query stack
error: aborting due to 2 previous errors; 1 warning emitted

For more information about this error, try `rustc --explain E0277`.
```

</details>

--- a/compiler/rustc_mir_transform/src/shim/async_destructor_ctor.rs
+++ b/compiler/rustc_mir_transform/src/shim/async_destructor_ctor.rs
@@ -6,7 +6,7 @@ use rustc_middle::mir::{
     BasicBlock, BasicBlockData, Body, Local, LocalDecl, MirSource, Operand, Place, Rvalue,
     SourceInfo, Statement, StatementKind, Terminator, TerminatorKind,
 };
-use rustc_middle::ty::{self, EarlyBinder, Ty, TyCtxt};
+use rustc_middle::ty::{self, EarlyBinder, Ty, TyCtxt, TypeVisitableExt};
 
 use super::*;
 use crate::patch::MirPatch;
@@ -121,9 +121,10 @@ pub(super) fn build_async_drop_shim<'tcx>(
         parent_args.as_coroutine().resume_ty(),
     )));
     body.phase = MirPhase::Runtime(RuntimePhase::Initial);
-    if !needs_async_drop {
+    if !needs_async_drop || drop_ty.references_error() {
         // Returning noop body for types without `need async drop`
-        // (or sync Drop in case of !`need async drop` && `need drop`)
+        // (or sync Drop in case of !`need async drop` && `need drop`).
+        // And also for error types.
         return body;
     }
 

--- a/tests/ui/async-await/async-drop/open-drop-error2.rs
+++ b/tests/ui/async-await/async-drop/open-drop-error2.rs
@@ -1,0 +1,21 @@
+//@compile-flags: -Zvalidate-mir -Zinline-mir=yes --crate-type=lib
+
+#![feature(async_drop)]
+#![allow(incomplete_features)]
+
+use std::{
+    future::{Future, async_drop_in_place},
+    pin::pin,
+    task::Context,
+};
+
+fn wrong() -> impl Sized {
+    //~^ ERROR: the size for values of type `str` cannot be known at compilation time
+    *"abc" // Doesn't implement Sized
+}
+fn weird(context: &mut Context<'_>) {
+    let mut e = wrong();
+    let h = unsafe { async_drop_in_place(&raw mut e) };
+    let i = pin!(h);
+    i.poll(context);
+}

--- a/tests/ui/async-await/async-drop/open-drop-error2.stderr
+++ b/tests/ui/async-await/async-drop/open-drop-error2.stderr
@@ -1,0 +1,19 @@
+error[E0277]: the size for values of type `str` cannot be known at compilation time
+  --> $DIR/open-drop-error2.rs:12:15
+   |
+LL | fn wrong() -> impl Sized {
+   |               ^^^^^^^^^^ doesn't have a size known at compile-time
+LL |
+LL |     *"abc" // Doesn't implement Sized
+   |     ------ return type was inferred to be `str` here
+   |
+   = help: the trait `Sized` is not implemented for `str`
+help: references are always `Sized`, even if they point to unsized data; consider not dereferencing the expression
+   |
+LL -     *"abc" // Doesn't implement Sized
+LL +     "abc" // Doesn't implement Sized
+   |
+
+error: aborting due to 1 previous error
+
+For more information about this error, try `rustc --explain E0277`.

